### PR TITLE
Ad hoc polymorphism

### PR DESCRIPTION
**Pull request by ryzhyk**
Implements #674 

Details - in commit messages.

Also adds support for recursive functions along the way.

Not in this PR:
- de-uglified string conversion mechanism.
- object-oriented function call syntax
- polymorphic aggregation functions

I will work on these next.

### CODE REVIEW

File: doc/tutorial/tutorial.md [hunk 1]
```diff
@@ -815,6 +842,23 @@ pub fn string_slice(x: &String, from: &u64, to: &u64) -> String {
...
+optimizations.  The programmer is responsible for labeling such functions with
+the `#[has_side_effects]` attribute, e.g., the following function is defined in
+the `log.dl` library:
+
+```
+#[has_side_effects]
+extern function log(module: module_t, level: log_level_t, msg: string): ()
+```
+
+The compiler automatically infers these annotations for non-extern functions, so
```
Comment (mbudiu-vmw) [inline doc/tutorial/tutorial.md:859]: This seems inconsistent, before you said that non-extern functions cannot have side-effects. So what annotations are inferred?

File: src/Language/DifferentialDatalog/Compile.hs [hunk 2]
```diff
@@ -2592,6 +2596,15 @@ mkExpr' d ctx EAs{..} | bothIntegers && narrow_from && narrow_to && width_cmp /=
 
 mkExpr' _ _ e = error $ "Compile.mkExpr': unexpected expression at " ++ show (pos e)
 
+mkFuncName :: DatalogProgram -> Function -> Doc
+mkFuncName d f | length namesakes == 1 = rname $ name f
+               | otherwise =
+    (rname $ name f) <> "_" <> targ0 <> "_" <> pp (length $ funcArgs f)
+    where
+    arg0 = funcArgs f !! 0
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Compile.hs:2604]: why can't this go out of bounds?

File: src/Language/DifferentialDatalog/Expr.hs [hunk 3]
```diff
@@ -433,18 +451,22 @@ exprIsVarOrField' _                = False
...
 -- No clever analysis here; just the obvious cases.
 exprIsInjective' :: DatalogProgram -> ExprNode Bool -> Bool
 exprIsInjective' _ EVar{}        = True
-exprIsInjective' d EApply{..}    =
-    -- FIXME: once we add support for recursive functions, be careful to avoid
-    -- infinite recursion.  The simple thing to do is just to return False for
-    -- recursive functions, as reasoning about them seems tricky otherwise.
-    and exprArgs && (maybe False (exprIsInjective d (CtxFunc f) (S.fromList $ map (\a -> ArgVar f $ name a) funcArgs)) $ funcDef)
-    where f@Function{..} = getFunc d exprFunc
+exprIsInjective' _ EApply{..}    = and exprArgs
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Expr.hs:469]: what happened to recursive functions?

File: src/Language/DifferentialDatalog/Type.hs [hunk 4]
```diff
@@ -257,10 +258,12 @@ exprNodeType' d ctx (EVar p v)            =
...
-exprNodeType' d ctx (EApply _ f _) | -- Type inference engine annotates calls to functions whose return type is polymorphic.
-                                     typeIsPolymorphic t  = ctxExpectType ctx
-                                   | otherwise            = t
-    where t = funcType $ getFunc d f
+exprNodeType' d ctx (EApply _ [f] ts) | -- Type inference engine annotates calls to functions whose return type is polymorphic.
+                                      typeIsPolymorphic t  = ctxExpectType ctx
+                                      | otherwise            = t
+    where t = funcType $ getFunc d f ts
+
+exprNodeType' _ _   e@EApply{} = error $ "exprNodeType' called with unresolved functio name: " ++ show e
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Type.hs:266]: typo in error message

File: src/Language/DifferentialDatalog/Validate.hs [hunk 5]
```diff
@@ -80,22 +81,27 @@ validate d = do
...
     -- This check must be done after 'depGraphValidate', which may
     -- introduce recursion
     checkNoRecursion d''
+    -}
     -- Attributes do not affect the semantics of the program and can therefore
     -- be validated last.
     progValidateAttributes d''
     return d''
 
+{-
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Validate.hs:94]: are you still doing this someplace else?

File: test/antrea/networkpolicy_controller.dl [hunk 6]
```diff
@@ -558,12 +558,12 @@ AppliedToGroupPod(appliedToGroup, k8s.PodReference{pod.name, pod.namespace}, pod
...
 AppliedToGroupPodsByNode(appliedToGroup, nodeName, pods) :-
     AppliedToGroupPod(appliedToGroup, pod, nodeName),
-    var pods = Aggregate((appliedToGroup, nodeName), group2set(pod)).
+    var pods = Aggregate((appliedToGroup, nodeName), group_to_set(pod)).
     //var podsByNode = Aggregate((appliedToGroup), group2map((nodeName, podsOnNode))).
 
 AppliedToGroupSpan(appliedToGroup, nodeName) :-
     AppliedToGroupPod(appliedToGroup, _, nodeName).
-//    var span = Aggregate((appliedToGroup), group2set(nodeName)).
+//    var span = Aggregate((appliedToGroup), group`/2set(nodeName)).
```
Comment (mbudiu-vmw) [inline test/antrea/networkpolicy_controller.dl:566]: typo in comment?

File: doc/tutorial/tutorial.md (hunk 1 above)
Comment (ryzhyk) [reply doc/tutorial/tutorial.md:859]: non-extern functions can invoke extern functions with side effects.  I will clarify.

File: src/Language/DifferentialDatalog/Compile.hs (hunk 2 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Compile.hs:2604]: `arg0` is only evaluated if this is a polymorphic function.  Polymorphic functions must take at least one argument (enforced in Validate.hs)

File: src/Language/DifferentialDatalog/Expr.hs (hunk 3 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Expr.hs:469]: oops. I actually implemented support for them but forgot to use it here.

File: src/Language/DifferentialDatalog/Validate.hs (hunk 5 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Validate.hs:94]: no, recursive functions are now allowed

File: src/Language/DifferentialDatalog/Expr.hs (hunk 3 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Expr.hs:469]: no, we're good, that check just moved above.

**Pull request merged**